### PR TITLE
For #25810 - Unified search 'search popup menu' should give hint to the user that there are more menu items available by scrolling

### PR DESCRIPTION
**Pull request by t-p-white**
Fenix changes here: https://github.com/mozilla-mobile/fenix/pull/27895

The unified search 'search popup menu' should give hint to the user that there are more menu items available by scrolling. This is indicated by displaying a half menu item if the available height will not fit the full list of items.

Existing behaviour:

https://user-images.githubusercontent.com/1833156/202675300-ab202c8f-5d3c-4fa9-aed5-bca265d872d7.mov

New behaviour LTR: 

https://user-images.githubusercontent.com/1833156/210393163-c7e3aece-c3da-4ccd-8eef-b221d7821ba3.mov

New behaviour RTL: 

https://user-images.githubusercontent.com/1833156/210393280-c810b7a4-b8e9-4d80-988c-150c4c69e01c.mov

### UI states:
* Does menu fit or does menu have to scroll
* Search bar Top/Bottom
  * Portrait keyboard open
  * Portrait keyboard closed
  * Landscape keyboard open
  * Landscape keyboard closed

**Note**: This issue is covering the direction and popup height only, another issue is in progress for positioning https://github.com/mozilla-mobile/fenix/issues/27998.

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation

Used by GitHub Actions.

### CODE REVIEW

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/BrowserMenuController.kt
Comment (t-p-white) [inline android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/BrowserMenuController.kt:57]: Note: This is also consistent with the 'settings' menu popup

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (t-p-white) [inline android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:80]: Adding logging as a starter - should this do something else to capture the error e.g. included in telemetry data?

**Comment by mavduevskiy** [comment]
@t-p-white 
just a heads up: for a successful build `BrowserMenuControllerTest` needs to be updated🙃

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/BrowserMenuController.kt
Comment (mavduevskiy) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/BrowserMenuController.kt:57]: @rocketsroger 
I found a solid [reasoning](https://github.com/mozilla-mobile/fenix/pull/17682#issuecomment-769883653) why the hosts (fragments, controllers) handle the dismissal of the menus (like [here](https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt#1475)).  

I was wondering why the same approach hasn't been used towards bookmarks fragment or tabs tray? They keep the `BrowserMenu` at the same position, and Bookmarks fragment actually chops a good chunk from the menu list. Could it be a universal approach that we want to hide menus after the configuration change because of misalignment?

Comment (rocketsroger) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/BrowserMenuController.kt:57]: As discussed with @mavduevskiy, I think in this case we're not using `OnConfigurationChanged` but using layout to determine if we should dismiss the menu.  However, we could add a parameter in `show()` to allow the user to determine that it should "auto dismiss"?

File: android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt
Comment (mavduevskiy) [inline android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt:240]: magic number with no comments :) 
50 / 8 = 6 – item height
40 anchor height, so 6 items could fit. we crop the last one, so it's 33, plus 1 padding == 34!

File: android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt
Comment (mavduevskiy) [inline android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt:259]: I wonder if there is a way not to hardcore the expected height? At least an explanation of the math, it required debugging to understand where the expected value comes from.
The calculation is something like this
`val expectedMenuContainerHeight = (itemHeight * ((viewHeight / itemHeight) - 0.5)).toInt() + viewPadding`

**Comment by Verdi** [comment]
After discussing this we want the menu to:

- Take up as much vertical space as it needs to show the entire list.
  - If we can't display the whole list it should be truncated in the middle of an item to indicate that it's scrollable
- If the toolbar is on bottom the menu should be aligned with the bottom of the viewport.
- If the toolbar is on top the menu should be aligned with the top of the viewport.

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (t-p-white) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:21]: I just found it a more usable name while doing dev but happy to revert as it was 👍

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (t-p-white) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:46]: Good observation! It was intentional, however 6 may be more optimal

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (mavduevskiy) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:21]: Personally, I might be finding it more usable as well, but when introducing changes to the already excising code we should consider if we follow (or break) the pattern that is already existing and if there is a reason for that and if the change if really necessary. Here, the same object is being passed down and it didn't change its parameters or behavior, so changing its name seems not necessary (though I do agree that `menuViewContainer` is a more accurate naming!)

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (mavduevskiy) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:19]: It might be a good idea to mention such "behind the scene" discussions briefly in PR or add a comment around that change, to prevent any questions

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/BrowserMenuController.kt
Comment (t-p-white) [inline android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/BrowserMenuController.kt:44]: Is there a reason for orientation being nullable? To me, it seems like this could cause unpredictable behaviour or a bug.

**Comment by mavduevskiy** [comment]
For the future reference: one of the approaches we were trying here was to stop relying on the `showAsDropDown` method and use only `showAtLocation`, but decided against it. 

Going through the source code one could find that the main difference between the two is that `showAsDropDown` will reposition itself after layout change while `showAtLocation` won't. Also, `showAsDropDown` is aligned to the anchor, while `showAtLocation` will be displayed at give `x, y` coordinates, which does mean completely manual position calculation.

**Comment by mavduevskiy** [comment]
Heading into right direction!

A few notes:
1. `numberOfItemsFitExactly` could be improved to account for rounding – if we can fit 6.9 items, we will display just 5.5. Displaying 6.5 sounds more reasonable. Not critical.
3. The case with the keyboard and showing the menu from the bottom could be handled by something like this:
```
orientation == Orientation.UP && containerViewHeight > availableHeightToTop ->
    showPopupWhereBestFits(
        anchor,
        fitsUp,
        fitsDown,
        isRtl,
        containerHeight,
    )
```
`showAtLocation` in this case should switch to using `anchor.getLocationInWindow(anchorPosition)` for getting the correct coordinates in the landscape mode of the `searchFragment`.
4. `showAtAnchorLocation` should also apply the possible shortened height of the view.
5. ktlint fails!

**Comment by t-p-white** [comment]
Thank you @mavduevskiy, really useful feedback. The above comments should now be resolved. As discussed, we'll hold off landing this until @rocketsroger has had a chance to also take a loo .

File: android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt
Comment (mavduevskiy) [inline android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt:130]: Maybe we could change the ambiguous `called with expected config` THEN condition to something more specific? Like 
`WHEN orientation up & does fit available height THEN the popup keeps its height & is positioned from the bottom of the anchor with leftBottomAnimation`

It is a mouthful, but it describes the business requirement we have. It might come in handy if we change the behavior: tests will start to fail, and the name of the test would tell us exactly what is the use case we should fix.

File: android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt
Comment (mavduevskiy) [inline android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt:280]: easy to read, nice!

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/BrowserMenuController.kt
Comment (mavduevskiy) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/BrowserMenuController.kt:42]: Looking at [the patch](https://github.com/rocketsroger/android-components/commit/87c168eab46ea1570aeabe029e48af3b18331e39) I don't see a reason why it should be null. I think we could make it not nullable as a follow up. Pinging the author, maybe there was a not-so-obvious reason
 @rocketsroger

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (mavduevskiy) [inline android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:211]: Nit: same comment to a single line function as before, I find it to be an overkill for what it hides :)

 ```
   val numberOfItemsFitWithOverflow = numberOfItemsFitExactly - HALF_MENU_ITEM
```

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (mavduevskiy) [inline android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:218]: single line

**Review by mavduevskiy — CHANGES_REQUESTED** [review]
Looking good! Few small nits here and there.

The main things are: isRTL vs reversed, improving the naming of tests to explain clearly the condition we are trying to match, reducing extra lines for improved readability.

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (t-p-white) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:150]: Thanks!

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (t-p-white) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:238]: Good spot

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (t-p-white) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:168]: That's fair. It was doing more prior to the latest change, but agree that it is overkill now

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (t-p-white) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:78]: My thoughts are that it must not have been the correct to begin with. Unless I’ve misunderstood, the original `val reversed = availableWidthToLeft < availableWidthToRight`  would have actually been true if it was LTR?
image

File: android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt
Comment (t-p-white) [reply android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt:64]: Good suggestion. I've updated to boolean as a starter - this may change based on feedback from on @rocketsroger around the displayPopup function error handling

**Review by mavduevskiy — CHANGES_REQUESTED** [review]
Looks good to me! One thing before approval, maybe we should change the way RTL is determined? Currently it relies heavily on the implied knowledge about the anchor position (at start), and won't work correctly for the icon positioned at the end. We could improve that by getting orientation from the anchor itself.

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (t-p-white) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:103]: > Maybe using view.isRTL inside PopupWindow.displayPopup would be a better approach? It would actually tell the direction of the layout without relying on the start/end position of the anchor

Good suggestion, I don't see any issue with this (@rocketsroger thoughts?) I'll update to use the view.

**Comment by t-p-white** [comment]
@mavduevskiy, @rocketsroger following demo of the updated behaviour, @Verdi approves 👍

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (mavduevskiy) [inline android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:84]: Looking at the name of the parameter and the name of the function `getContainerHeight` and the name of passed parameter `containerViewHeight` to be a bit confusing.  `containerViewHeight` is the height of the container if it's drawn as it is, but then the function is calculating the new height. Maybe changing the name from `getContainerHeight` (since we already have it) to `getNewContainerHeight` or `calculateContainerHeight` would be a more clear way to tell what's happening between a passed height and the new calculated height.

**Review by mavduevskiy — APPROVED** [review]
One naming nit (optional, change if you like the suggestion) and let's ship it! I really like the feature, good job! 

I would prefer @rocketsroger to take a final look before we ship it.

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (rocketsroger) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:57]: Is this something that you expect might happen in a corner case? Or is this just something you want to record in case that it does happen?

**Review by rocketsroger — APPROVED** [review]
Looks good to me.  Only question left is the error when view is initialized incorrectly and the comment about naming @mavduevskiy asked.

File: android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt
Comment (t-p-white) [reply android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt:57]: Definitely a corner case - I can't think of a situation where this could happen as we always have the default menu items (which can not be removed by the user). If this does somehow happen then something very unexpected has occurred. I think by checking the conditions, we can at least provide somewhat of a graceful experience - the logging may be superfluous if it is not accessible (would we get sight of errors or would this be local to the device?)

**Pull request merged**
